### ISSUE TITLE
Role dependency, failed to template: recursive loop detected in template string

### ISSUE THREAD

**Issue by hryamzik** [ansible/ansible#12713]
##### Issue Type:
- Bug Report
##### Ansible Version:

1.9.3
##### Ansible Configuration:

Default.
##### Environment:

Mac OS 10.10.5, applied on Ubuntu Linux 14.04
##### Summary:

Promoting role defaults to dependent roles leads to an error.

Let's say you have a `starman` role with a default variable `number_of_starman_workers` set to `100`.
Now you create another role, `my_app`, with a default value of `number_of_starman_workers` set to `10`. `my_app` depends on `starman` but in a starman role `number_of_starman_workers` turns out to be `100`. So to promote defaults from `my_app` you have to parametrise dependent role, and here we fail.

The workaround is to give this variable a different name on `my_app` role, like `my_app_number_of_starman_workers`.
##### Steps To Reproduce:

Create two roles, `a` and `b`. Set default variable `var` in role `a`:

a/defaults/main.yml:

``` yaml
var: 4
```

And set dependency for role `b`L

a/meta/main.yml:

``` yml
dependencies:
  - role: b
    var: "{{ var }}"
```
##### Expected Results:

Variable `var` set.
##### Actual Results:

```
 Failed to template {{ var }}: recursive loop detected in template string: {{ var }}
```

**Comment by bcoca** [comment]
This is not really a bug, a var cannot be recursively templated to itself and vars should be passed trough to dependent roles anyways so there is not need for this either.

**Comment by hryamzik** [comment]
@bcoca in this case var is not set to itself!

**Comment by derekmahar** [comment]
I think this is still an issue.  I encountered a case similar to this and [_ansible: recursive loop detected in template string_](http://stackoverflow.com/questions/21738661/ansible-recursive-loop-detected-in-template-string).  I commented that I expected nested role invocation to work similar to nested function calls.  Essentially the problem happens because the parser doesn't treat each role as a separate variable scope.  For example, in the case of `{ role: project, project_name: "{{project_name}}" }`, the parser should first resolve `{{ project_name }}` to a value and then assign that value to variable `project_name`, but in the context of role `project`.  Though the variables have the same identifier, they are actually in different contexts.  This is exactly how function parameters work.
